File: src/restaurantChain/Restaurant.java
package restaurantChain;

import java.io.ObjectOutputStream.PutField;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class Restaurant {

	private String name;
	private int tables, busyTables = 0, refused = 0;
	private Map<String, Menu> listing = new HashMap<>();
	private Map<String, Integer> prenotations = new HashMap<String, Integer>();
	private Map<String, String[]> orders = new HashMap<>(); 
	private Map<String, Double> paid = new HashMap<String, Double>();

	public Restaurant(String name, int tables) {
		this.name = name;
		this.tables = tables;
	}

	public String getName() {
		return name;
	}

	public void addMenu(String name, double price) throws InvalidName {
		if (listing.containsKey(name)) {
			throw new InvalidName();
		}
		listing.put(name, new Menu(name, price));
	}

	public int reserve(String name, int persons) throws InvalidName {
		if (prenotations.containsKey(name))
			throw new InvalidName();
		int requiredTables = (int) Math.ceil(persons / 4);
		if (tables - busyTables - requiredTables < 0) {
			refused ++;
			return 0;
		}
		busyTables += requiredTables;
		prenotations.put(name, persons);
		return requiredTables;
	}

	public int getRefused() {
		return refused;
	}

	public int getUnusedTables() {
		return tables - busyTables;
	}

	public boolean order(String name, String... menu) throws InvalidName {
		if (!prenotations.containsKey(name)) {	throw new InvalidName();	}
		if (menu.length< prenotations.get(name)) return false;
		for (String m : menu) {
			if(!listing.containsKey(m)) throw new InvalidName();
		}
		orders.put(name, menu);
		return true;
	}

	public List<String> getUnordered() {
		
		List<String> unordered = new ArrayList<String>();
		prenotations.keySet().forEach(p -> {
			if (!orders.containsKey(p)) {
				unordered.add(p);
			}
		});
		return unordered;
	}

	public double pay(String name) throws InvalidName {
		if (!prenotations.containsKey(name)) throw new InvalidName();
		if (!orders.containsKey(name)) throw new InvalidName();
		double toPay = 0;
		for ( String m : orders.get(name)) {
			toPay += listing.get(m).getPrice();
		}
		paid.put(name, toPay);
		return toPay;
	}

	public List<String> getUnpaid() {
		List<String> unpaid = new  ArrayList<String>();
		orders.keySet().forEach( o -> {
			if(!paid.containsKey(o)){
				unpaid.add(o);
			}
		} );
		return unpaid;
	}

	public double getIncome() {
		return paid.values().stream().reduce(0.0, (x,y) -> x+y );
	}

}

File: src/restaurantChain/Restaurant.kt
 package restaurantChain

 import java.util.stream.Collectors

 class Restaurant(val name: String?, val nTables: Int) {

    val listing = mutableMapOf<String, Menu>()
    val prenotations = mutableMapOf<String, Int>()
    var busyTables: Int = 0
    var refused: Int = 0
    val orders = mutableMapOf<String, List<String?>>()
    val paid = mutableMapOf<String, Double>()

     @Throws (InvalidName::class)
    fun addMenu(name: String, price: Double) {
        if(listing.containsKey(name)){
            throw InvalidName()
        }
        listing.put(name , Menu(name, price))
    }

     @Throws (InvalidName::class)
     fun reserve(name: String, persons: Int): Int{
        if (prenotations.containsKey(name))
            throw InvalidName()
        val requiredTables = Math.ceil(((persons.toDouble()/4)))
        if (nTables - requiredTables - busyTables < 0){
            refused += persons
            return 0
        }
        busyTables += requiredTables.toInt()
        prenotations.put(name, persons)
        return requiredTables.toInt()
    }

    fun getUnusedTables() =  nTables - busyTables

     @Throws (InvalidName::class)
    fun order(name: String?, vararg menu : String? ): Boolean{
        if (!prenotations.containsKey(name))
            throw InvalidName()
        if (menu.size < prenotations[name] as Int)
            return false
        for (m in menu){
            if (!listing.containsKey(m))
                throw InvalidName()
        }
        if (name != null) {
            orders.put(name, menu.toList())
        }
        return true
    }

    fun getUnordered(): List<String>{
        return prenotations.keys.stream().filter({ o -> !orders.containsKey(o) }).sorted().collect(Collectors.toList())
    }

     @Throws (InvalidName::class)
    fun pay(name: String): Double{
        if (!prenotations.containsKey(name)) throw InvalidName()
        if (!orders.containsKey(name)) return 0.0
        var toPay = 0.0
        for (m in orders.get(name) as List){
            toPay += listing.get(m)?.price as Double
        }
        paid.put(name, toPay)
        return toPay
    }

     fun getUnpaid(): List<String>{
         return  orders.keys.filter { o -> !paid.containsKey(o) }.sorted().toList()
     }

     fun getIncome(): Double{
         return paid.values.stream().reduce(0.0, {x,y -> x+y})
     }
}